File: tests/test_api.py
"""Comprehensive Automated Tests for Smart Water Tank System (SRS Compliance)."""

import os
import tempfile
import pytest
from fastapi.testclient import TestClient

# Use temporary database for test isolation
temp_db = tempfile.NamedTemporaryFile(suffix=".db", delete=False)
temp_db.close()
os.environ["TANK_DB_PATH"] = temp_db.name

from backend.database import init_db, get_db_connection
from backend.main import app

client = TestClient(app)


@pytest.fixture(scope="session", autouse=True)
def setup_test_db():
    """Initialize test database before running tests."""
    init_db()
    yield
    try:
        os.remove(temp_db.name)
    except Exception:
        pass


@pytest.fixture
def admin_token():
    """Login as admin and return JWT token."""
    response = client.post("/api/auth/login", json={"username": "admin", "password": "admin123"})
    assert response.status_code == 200
    return response.json()["access_token"]


@pytest.fixture
def operator_token():
    """Login as operator and return JWT token."""
    response = client.post("/api/auth/login", json={"username": "operator", "password": "operator123"})
    assert response.status_code == 200
    return response.json()["access_token"]


# =====================================================================
# 1. Health & System Tests
# =====================================================================

def test_health_check():
    response = client.get("/api/health")
    assert response.status_code == 200
    assert response.json()["status"] == "healthy"


# =====================================================================
# 2. Authentication Tests (NFR-03)
# =====================================================================

def test_login_success():
    response = client.post("/api/auth/login", json={"username": "admin", "password": "admin123"})
    assert response.status_code == 200
    data = response.json()
    assert "access_token" in data
    assert data["user"]["username"] == "admin"
    assert data["user"]["role"] == "admin"


def test_login_invalid_password():
    response = client.post("/api/auth/login", json={"username": "admin", "password": "wrongpassword"})
    assert response.status_code == 401
    assert "Invalid username or password" in response.json()["detail"]


def test_login_nonexistent_user():
    response = client.post("/api/auth/login", json={"username": "ghost_user", "password": "password"})
    assert response.status_code == 401


def test_get_current_user_me(admin_token):
    response = client.get("/api/auth/me", headers={"Authorization": f"Bearer {admin_token}"})
    assert response.status_code == 200
    assert response.json()["username"] == "admin"


def test_get_current_user_unauthorized():
    response = client.get("/api/auth/me")
    assert response.status_code == 401


# =====================================================================
# 3. Safe Sediment Limit Tests (FR-03, NFR-03)
# =====================================================================

def test_get_safe_limit_public():
    """Any user can view the configured limit."""
    response = client.get("/api/limit")
    assert response.status_code == 200
    assert "safe_limit" in response.json()


def test_update_safe_limit_unauthorized():
    """NFR-03: Setting safe limit without credentials MUST fail with 401."""
    response = client.put("/api/limit", json={"safe_limit": 65.0})
    assert response.status_code == 401


def test_update_safe_limit_authorized(admin_token):
    """FR-03: Authorized user can set safe sediment level limit."""
    response = client.put(
        "/api/limit",
        json={"safe_limit": 65.0},
        headers={"Authorization": f"Bearer {admin_token}"}
    )
    assert response.status_code == 200
    assert response.json()["safe_limit"] == 65.0
    assert response.json()["updated_by"] == "admin"


def test_update_safe_limit_validation(admin_token):
    """Validation: limit must be between 1 and 100."""
    # Under min
    response = client.put(
        "/api/limit",
        json={"safe_limit": 0.5},
        headers={"Authorization": f"Bearer {admin_token}"}
    )
    assert response.status_code == 422

    # Over max
    response = client.put(
        "/api/limit",
        json={"safe_limit": 105.0},
        headers={"Authorization": f"Bearer {admin_token}"}
    )
    assert response.status_code == 422


# =====================================================================
# 4. Sensor Reading Ingestion & Automatic Control (FR-01, FR-04, FR-05, FR-06)
# =====================================================================

def test_ingest_normal_sensor_reading():
    """Normal reading records to SQLite and maintains supply ON."""
    response = client.post("/api/sensor/readings", json={"sediment_level": 30.0})
    assert response.status_code == 201
    data = response.json()
    assert data["sediment_level"] == 30.0
    assert data["status"] == "NORMAL"
    assert data["supply_status"] == "ON"


def test_automatic_alert_and_supply_shutoff_when_exceeded():
    """
    FR-04: System shall send an alert when sediment level exceeds limit.
    FR-05: System shall automatically stop water supply when sediment exceeds limit.
    FR-06: Store sensor reading, alert, and supply status in SQLite.
    """
    # Set limit to 60.0
    client.put("/api/limit", json={"safe_limit": 60.0}, headers={"Authorization": f"Bearer {client.post('/api/auth/login', json={'username': 'admin', 'password': 'admin123'}).json()['access_token']}"})

    # Ingest reading of 78.5% (exceeds 60.0%)
    response = client.post("/api/sensor/readings", json={"sediment_level": 78.5})
    assert response.status_code == 201
    data = response.json()
    assert data["status"] == "CRITICAL"
    assert data["supply_status"] == "OFF"

    # Verify Alert was automatically created in SQLite (FR-04, NFR-02)
    alert_resp = client.get("/api/alerts")
    assert alert_resp.status_code == 200
    alerts = alert_resp.json()["items"]
    assert len(alerts) > 0
    latest_alert = alerts[0]
    assert latest_alert["severity"] == "CRITICAL"
    assert "exceeds safe limit" in latest_alert["message"]

    # Verify Water Supply is OFF (FR-05)
    supply_resp = client.get("/api/supply/status")
    assert supply_resp.status_code == 200
    assert supply_resp.json()["status"] == "OFF"

    # Verify Supply log entry in SQLite (FR-06)
    logs_resp = client.get("/api/supply/logs")
    assert logs_resp.status_code == 200
    assert logs_resp.json()["total"] > 0
    latest_log = logs_resp.json()["items"][0]
    assert latest_log["status"] == "OFF"
    assert "Automatic shutoff" in latest_log["reason"]


def test_automatic_water_supply_restore_when_sediment_normal():
    """When sediment returns below safe limit, water supply automatically resumes."""
    # Submit safe reading 35.0%
    response = client.post("/api/sensor/readings", json={"sediment_level": 35.0})
    assert response.status_code == 201
    assert response.json()["status"] == "NORMAL"
    assert response.json()["supply_status"] == "ON"

    # Verify Water Supply is ON
    supply_resp = client.get("/api/supply/status")
    assert supply_resp.json()["status"] == "ON"


# =====================================================================
# 5. Dashboard Real-Time Status Endpoint (FR-02, NFR-01, NFR-04)
# =====================================================================

def test_dashboard_latest_status():
    response = client.get("/api/sensor/latest")
    assert response.status_code == 200
    data = response.json()
    assert "current_sediment" in data
    assert "safe_limit" in data
    assert "status" in data
    assert "supply_status" in data
    assert "active_alerts_count" in data


# =====================================================================
# 6. CRUD Operations (Search, Filter, Acknowledge, Delete)
# =====================================================================

def test_sensor_readings_search_and_filter():
    response = client.get("/api/sensor/readings?status=CRITICAL")
    assert response.status_code == 200
    items = response.json()["items"]
    for item in items:
        assert item["status"] == "CRITICAL"


def test_delete_reading(admin_token):
    # Create reading to delete
    create_resp = client.post("/api/sensor/readings", json={"sediment_level": 44.0})
    reading_id = create_resp.json()["id"]

    # Delete with authorization
    del_resp = client.delete(f"/api/sensor/readings/{reading_id}", headers={"Authorization": f"Bearer {admin_token}"})
    assert del_resp.status_code == 200


def test_acknowledge_alert(operator_token):
    # Trigger alert
    client.post("/api/sensor/readings", json={"sediment_level": 92.0})
    alerts = client.get("/api/alerts").json()["items"]
    alert_id = alerts[0]["id"]

    # Acknowledge
    ack_resp = client.post(f"/api/alerts/{alert_id}/ack", headers={"Authorization": f"Bearer {operator_token}"})
    assert ack_resp.status_code == 200

    # Verify acknowledged flag
    refreshed_alerts = client.get("/api/alerts").json()["items"]
    target = next(a for a in refreshed_alerts if a["id"] == alert_id)
    assert target["acknowledged"] is True
    assert target["acknowledged_by"] == "operator"


# =====================================================================
# 7. Reports & CSV Export Tests
# =====================================================================

def test_reports_summary():
    response = client.get("/api/reports/summary")
    assert response.status_code == 200
    data = response.json()
    assert data["total_readings"] > 0
    assert "average_sediment" in data
    assert "compliance_rate_percent" in data


def test_reports_csv_export():
    for r_type in ["readings", "alerts", "supply"]:
        response = client.get(f"/api/reports/export?report_type={r_type}")
        assert response.status_code == 200
        assert response.headers["content-type"].startswith("text/csv")
        assert len(response.text) > 0


# =====================================================================
# 8. Sensor Simulator Tests
# =====================================================================

def test_simulator_inject():
    response = client.post("/api/simulator/inject", json={"sediment_level": 52.5})
    assert response.status_code == 200
    assert response.json()["sediment_level"] == 52.5

File: tests/test_api.ts
/**
 * Automated System Verification Test Suite
 * Tests SRS Requirements: FR-01 through FR-06, NFR-01 through NFR-06
 * Tests Database Migrations, Relationships, Constraints, and Supabase / SQLite Parity
 */

const BASE_URL = process.env.TEST_BASE_URL || "http://127.0.0.1:3000";

let adminToken = "";
let operatorToken = "";
let createdReadingId: number | null = null;
let triggeredAlertId: number | null = null;

let testsPassed = 0;
let testsFailed = 0;

async function runTest(name: string, fn: () => Promise<void>) {
  try {
    await fn();
    console.log(`✅ [PASS] ${name}`);
    testsPassed++;
  } catch (err: any) {
    console.error(`❌ [FAIL] ${name}: ${err.message || err}`);
    testsFailed++;
  }
}

function assert(condition: boolean, message: string) {
  if (!condition) {
    throw new Error(message);
  }
}

async function startTestSuite() {
  console.log(`\n======================================================`);
  console.log(`🧪 Running SmartTank Verification Tests against ${BASE_URL}`);
  console.log(`======================================================\n`);

  // Test 1: Health & Database Health Check
  await runTest("Test 01: System & Database Health Check (/api/health)", async () => {
    const res = await fetch(`${BASE_URL}/api/health`);
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(data.status === "healthy", "Expected status to be healthy");
    assert(data.database && data.database.connected === true, "Database should be connected");
  });

  // Test 2: Ingest sensor reading within 0-100 (FR-01)
  await runTest("Test 02: Ingest valid sensor reading (FR-01)", async () => {
    const res = await fetch(`${BASE_URL}/api/sensor/readings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sediment_level: 48.5 })
    });
    assert(res.status === 201, `Expected 201, got ${res.status}`);
    const data = await res.json();
    assert(data.sediment_level === 48.5, "Sediment level should match 48.5");
    assert(data.status === "NORMAL" || data.status === "WARNING", "Status should be classified");
    createdReadingId = data.id;
  });

  // Test 3: Sensor reading validation (reject < 0 or > 100)
  await runTest("Test 03: Sensor reading validation bounds rejection", async () => {
    const resHigh = await fetch(`${BASE_URL}/api/sensor/readings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sediment_level: 150.0 })
    });
    assert(resHigh.status === 422, `Expected 422 for >100, got ${resHigh.status}`);

    const resLow = await fetch(`${BASE_URL}/api/sensor/readings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sediment_level: -5.0 })
    });
    assert(resLow.status === 422, `Expected 422 for <0, got ${resLow.status}`);
  });

  // Test 4: Dashboard latest metrics retrieval (FR-02)
  await runTest("Test 04: Dashboard live metrics retrieval (FR-02)", async () => {
    const res = await fetch(`${BASE_URL}/api/sensor/latest`);
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(typeof data.current_sediment === "number", "current_sediment must be numeric");
    assert(typeof data.safe_limit === "number", "safe_limit must be numeric");
    assert(data.supply_status === "ON" || data.supply_status === "OFF", "supply_status must be ON or OFF");
  });

  // Test 5: Query safe sediment limit (FR-03)
  await runTest("Test 05: Retrieve current safe sediment limit (FR-03)", async () => {
    const res = await fetch(`${BASE_URL}/api/limit`);
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(typeof data.safe_limit === "number" && data.safe_limit > 0, "safe_limit must be positive");
  });

  // Test 6: Non-authenticated user cannot update safe limit (NFR-03)
  await runTest("Test 06: Reject unauthenticated limit update (NFR-03)", async () => {
    const res = await fetch(`${BASE_URL}/api/limit`, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ safe_limit: 65.0 })
    });
    assert(res.status === 401, `Expected 401 Unauthorized, got ${res.status}`);
  });

  // Test 7: User Login & JWT Token issuance
  await runTest("Test 07: User authentication for admin & operator", async () => {
    // Admin login
    const adminRes = await fetch(`${BASE_URL}/api/auth/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username: "admin", password: "admin123" })
    });
    assert(adminRes.status === 200, "Admin login should succeed");
    const adminData = await adminRes.json();
    assert(Boolean(adminData.access_token), "Admin token must be returned");
    adminToken = adminData.access_token;

    // Operator login
    const opRes = await fetch(`${BASE_URL}/api/auth/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username: "operator", password: "operator123" })
    });
    assert(opRes.status === 200, "Operator login should succeed");
    const opData = await opRes.json();
    assert(Boolean(opData.access_token), "Operator token must be returned");
    operatorToken = opData.access_token;
  });

  // Test 8: Admin updates safe sediment limit (FR-03)
  await runTest("Test 08: Admin updates safe sediment limit (FR-03)", async () => {
    const res = await fetch(`${BASE_URL}/api/limit`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${adminToken}`
      },
      body: JSON.stringify({ safe_limit: 68.0 })
    });
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(data.safe_limit === 68.0, "safe_limit should be updated to 68.0");
  });

  // Test 9: Safe limit input validation (reject < 1.0 or > 100.0)
  await runTest("Test 09: Safe limit input bounds validation (1.0 to 100.0)", async () => {
    const resOver = await fetch(`${BASE_URL}/api/limit`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${adminToken}`
      },
      body: JSON.stringify({ safe_limit: 105.0 })
    });
    assert(resOver.status === 422, `Expected 422, got ${resOver.status}`);

    const resZero = await fetch(`${BASE_URL}/api/limit`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${adminToken}`
      },
      body: JSON.stringify({ safe_limit: 0.0 })
    });
    assert(resZero.status === 422, `Expected 422, got ${resZero.status}`);
  });

  // Test 10: Critical sediment reading automatically generates alert (FR-04, NFR-02)
  await runTest("Test 10: Critical sediment generates alert (FR-04, NFR-02)", async () => {
    const res = await fetch(`${BASE_URL}/api/sensor/readings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sediment_level: 85.0 })
    });
    assert(res.status === 201, `Expected 201, got ${res.status}`);
    const reading = await res.json();
    assert(reading.status === "CRITICAL", "Status must be CRITICAL for 85% sediment (> 68% limit)");

    const alertRes = await fetch(`${BASE_URL}/api/alerts?size=5`);
    const alertData = await alertRes.json();
    assert(alertData.total > 0, "Alerts list should contain at least one alert");
    const latestAlert = alertData.items[0];
    assert(latestAlert.severity === "CRITICAL", "Severity must be CRITICAL");
    triggeredAlertId = latestAlert.id;
  });

  // Test 11: Critical sediment reading automatically shuts off water supply (FR-05)
  await runTest("Test 11: Critical sediment automatically stops water supply (FR-05)", async () => {
    const res = await fetch(`${BASE_URL}/api/supply/status`);
    const data = await res.json();
    assert(data.status === "OFF", `Water supply must be OFF, but got ${data.status}`);
    assert(data.control_mode === "AUTOMATIC", "Control mode must remain AUTOMATIC");
  });

  // Test 12: Normal sediment reading restores water supply automatically (FR-05)
  await runTest("Test 12: Safe sediment restores water supply automatically (FR-05)", async () => {
    const res = await fetch(`${BASE_URL}/api/sensor/readings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sediment_level: 32.0 })
    });
    assert(res.status === 201, "Normal reading should be ingested");

    const supplyRes = await fetch(`${BASE_URL}/api/supply/status`);
    const supplyData = await supplyRes.json();
    assert(supplyData.status === "ON", `Water supply should be restored to ON, got ${supplyData.status}`);
  });

  // Test 13: Operator acknowledges alert (FR-04, RBAC)
  await runTest("Test 13: Operator acknowledges alert (FR-04, RBAC)", async () => {
    if (!triggeredAlertId) {
      throw new Error("No alert ID to acknowledge");
    }
    const res = await fetch(`${BASE_URL}/api/alerts/${triggeredAlertId}/ack`, {
      method: "POST",
      headers: { "Authorization": `Bearer ${operatorToken}` }
    });
    assert(res.status === 200, `Expected 200, got ${res.status}`);
  });

  // Test 14: Water supply logs stored with audit details (FR-06)
  await runTest("Test 14: Water supply logs stored and queryable (FR-06)", async () => {
    const res = await fetch(`${BASE_URL}/api/supply/logs?size=10`);
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(data.total >= 2, "Should have logged initial startup, shutoff, and restore events");
    assert(data.items.length > 0, "Items array should not be empty");
  });

  // Test 15: Admin manual supply override (FR-05)
  await runTest("Test 15: Authorized user manual supply override (FR-05)", async () => {
    const res = await fetch(`${BASE_URL}/api/supply/override`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${adminToken}`
      },
      body: JSON.stringify({ status: "OFF", reason: "Scheduled line maintenance" })
    });
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(data.status === "OFF", "Status should be updated to OFF");

    // Restore back to ON
    const restoreRes = await fetch(`${BASE_URL}/api/supply/override`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${adminToken}`
      },
      body: JSON.stringify({ status: "ON", reason: "Maintenance complete" })
    });
    assert(restoreRes.status === 200, "Restore should succeed");
  });

  // Test 16: Sensor readings search and pagination (FR-06)
  await runTest("Test 16: Sensor readings pagination and search (FR-06)", async () => {
    const res = await fetch(`${BASE_URL}/api/sensor/readings?page=1&size=3`);
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(data.items.length <= 3, "Page size should be respected");
    assert(data.total >= 3, "Total records should be >= 3");
  });

  // Test 17: Reports summary calculations (FR-06)
  await runTest("Test 17: Reports & Analytics operational summary", async () => {
    const res = await fetch(`${BASE_URL}/api/reports/summary`);
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(typeof data.total_readings === "number" && data.total_readings > 0, "total_readings must be positive");
    assert(typeof data.average_sediment === "number", "average_sediment must be numeric");
    assert(typeof data.compliance_rate_percent === "number", "compliance_rate_percent must be numeric");
  });

  // Test 18: RFC 4180 CSV export for readings, alerts, and supply history
  await runTest("Test 18: RFC 4180 CSV Export endpoints", async () => {
    const readingsCsvRes = await fetch(`${BASE_URL}/api/reports/export?report_type=readings`);
    assert(readingsCsvRes.status === 200, "Readings CSV should return 200");
    const readingsCsv = await readingsCsvRes.text();
    assert(readingsCsv.includes("Sediment Level"), "CSV header should include Sediment Level");

    const alertsCsvRes = await fetch(`${BASE_URL}/api/reports/export?report_type=alerts`);
    assert(alertsCsvRes.status === 200, "Alerts CSV should return 200");
    const alertsCsv = await alertsCsvRes.text();
    assert(alertsCsv.includes("Severity"), "Alerts CSV should include Severity header");

    const supplyCsvRes = await fetch(`${BASE_URL}/api/reports/export?report_type=supply`);
    assert(supplyCsvRes.status === 200, "Supply CSV should return 200");
    const supplyCsv = await supplyCsvRes.text();
    assert(supplyCsv.includes("Water Supply Status"), "Supply CSV should include Water Supply Status header");
  });

  // Test 19: Sensor Simulator Toggle and Inject
  await runTest("Test 19: Sensor simulator control and direct injection", async () => {
    const statusRes = await fetch(`${BASE_URL}/api/simulator/status`);
    assert(statusRes.status === 200, "Simulator status should return 200");

    const injectRes = await fetch(`${BASE_URL}/api/simulator/inject`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ sediment_level: 52.4 })
    });
    assert(injectRes.status === 200, "Simulator inject should return 200");
    const injectData = await injectRes.json();
    assert(injectData.sediment_level === 52.4, "Injected sediment level should be 52.4");
  });

  // Test 20: CRUD Deletion RBAC & Record Maintenance
  await runTest("Test 20: CRUD Deletion RBAC & Record Maintenance", async () => {
    if (createdReadingId) {
      const delRes = await fetch(`${BASE_URL}/api/sensor/readings/${createdReadingId}`, {
        method: "DELETE",
        headers: { "Authorization": `Bearer ${adminToken}` }
      });
      assert(delRes.status === 200, "Admin can delete single reading");
    }

    // Non-admin cannot clear all readings
    const clearUnauthorized = await fetch(`${BASE_URL}/api/sensor/readings`, {
      method: "DELETE",
      headers: { "Authorization": `Bearer ${operatorToken}` }
    });
    assert(clearUnauthorized.status === 403, "Operator forbidden from clearing all readings");
  });

  // Test 21: AI Operational Diagnostic Report Generation
  await runTest("Test 21: AI Operational Diagnostic Report (/api/ai/diagnostics)", async () => {
    const res = await fetch(`${BASE_URL}/api/ai/diagnostics`, {
      method: "POST",
      headers: { "Content-Type": "application/json" }
    });
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(typeof data.executive_summary === "string" && data.executive_summary.length > 0, "Executive summary should be non-empty string");
    assert(["LOW", "MODERATE", "HIGH", "CRITICAL"].includes(data.risk_level), `Risk level should be valid enum, got ${data.risk_level}`);
    assert(typeof data.risk_score === "number" && data.risk_score >= 0 && data.risk_score <= 100, `Risk score should be 0-100, got ${data.risk_score}`);
    assert(typeof data.sediment_trend === "string" && data.sediment_trend.length > 0, "Sediment trend should be non-empty");
    assert(typeof data.supply_safety_assessment === "string" && data.supply_safety_assessment.length > 0, "Supply safety assessment should be non-empty");
    assert(Array.isArray(data.actionable_recommendations) && data.actionable_recommendations.length > 0, "Actionable recommendations should be non-empty array");
    assert(typeof data.generated_at === "string", "Generated timestamp should be present");
  });

  // Test 22: AI Operator Inquiry Endpoint
  await runTest("Test 22: AI Operator Inquiry (/api/ai/inquiry)", async () => {
    // Empty query test (should reject with 400)
    const emptyRes = await fetch(`${BASE_URL}/api/ai/inquiry`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query: "" })
    });
    assert(emptyRes.status === 400, `Expected 400 for empty query, got ${emptyRes.status}`);

    // Valid query test
    const res = await fetch(`${BASE_URL}/api/ai/inquiry`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ query: "Analyze current contamination risk and safe limit headroom." })
    });
    assert(res.status === 200, `Expected 200, got ${res.status}`);
    const data = await res.json();
    assert(typeof data.answer === "string" && data.answer.length > 0, "Answer should be non-empty string");
    assert(typeof data.source === "string", "Source should be present");
    assert(data.query === "Analyze current contamination risk and safe limit headroom.", "Query should match input");
  });

  console.log(`\n======================================================`);
  console.log(`📊 Test Results: ${testsPassed} passed, ${testsFailed} failed`);
  console.log(`======================================================\n`);

  if (testsFailed > 0) {
    process.exit(1);
  }
  process.exit(0);
}

startTestSuite().catch(err => {
  console.error("Test suite fatal error:", err);
  process.exit(1);
});
